ibis/backends/__init__.py

```diff
@@ -62,29 +87,70 @@ def __getattr__(self, name) -> ir.Table:
             raise AttributeError(name) from exc
 
     def __iter__(self) -> Iterator[str]:
-        return iter(sorted(self._backend.list_tables()))
+        return iter(sorted(self._tables))
 
     def __len__(self) -> int:
-        return len(self._backend.list_tables())
+        return len(self._tables)
 
     def __dir__(self) -> list[str]:
         o = set()
         o.update(dir(type(self)))
         o.update(
             name
-            for name in self._backend.list_tables()
+            for name in self._tables
             if name.isidentifier() and not keyword.iskeyword(name)
         )
         return list(o)
 
     def __repr__(self) -> str:
-        tables = self._backend.list_tables()
         rows = ["Tables", "------"]
-        rows.extend(f"- {name}" for name in sorted(tables))
+        rows.extend(f"- {name}" for name in sorted(self._tables))
         return "\n".join(rows)
 
     def _ipython_key_completions_(self) -> list[str]:
-        return self._backend.list_tables()
+        return self._tables
+
+
+class DDLAccessor:
+    """ddl accessor list views."""
+
+    def __init__(self, backend: BaseBackend):
+        self._backend = backend
```

Comment:
I know we have this elsewhere, but I think the circular reference here adds a lot of complexity to the lifetime semantics of a backend instance, especially due to the strong reference held by the `tables` `cached_property`.

Can we make this a weak reference? This ensures that a backend instance cannot stick around solely because there's a DDL accessor hanging around retaining a reference to the backend.

Reply:
I've only seen examples of us using weak references in some of our test. just to make sure, are you suggesting doing this? 

```python
    def __init__(self, backend: BaseBackend):
        self._backend = weakref.ref(backend)
```

Reply:
Thinking about this some more, I think we probably need to keep the reference cycle.

For example,

the following code should work:

```python
con = ibis.backend.connect(...)
tables = con.tables
del con
list(tables)
```

I think this would fail if `con` is a weakref.

Anyway, let's keep it as is until I can understand whether this would fail and why.

Reply:
Actually, having a weak referent it doesn't even let access the tables. It fails when I'm trying to do access teh methods to list things



```python
In [1]: import ibis

In [2]: con = ibis.get_backend()

In [3]: penguins = ibis.examples.penguins.fetch()

In [4]: con.tables
```

```python-traceback
File ~/Documents/git/my_forks/ibis/ibis/backends/__init__.py:122, in DDLAccessor._raise_if_not_implemented(self, method_name)
    121 def _raise_if_not_implemented(self, method_name: str):
--> 122     method = getattr(self._backend, method_name)
    123     if not callable(method):
    124         raise NotImplementedError(
    125             f"The method {method_name} is not implemented for the {self._backend.name} backend"
    126         )

AttributeError: 'weakref.ReferenceType' object has no attribute '_list_tables'
```

Reply:
That's because you're using `weakref.ref` incorrectly, deferencing requires invoking the object like a function:

```python
obj = my_object()
obj_ref = weakref.ref(obj)
underlying = obj_ref()
# `underlying` is either `None` or the referrent. 
```

What you want is `weakref.proxy(self)` which behaves like the object it holds.

I opened a PR to clean up existing circular references in https://github.com/ibis-project/ibis/pull/9895, you can see the

Reply:
I rebased (and will do if there are more changes) against your PR, to keep track of this plus my changes to be consistent.